#### About: 

We learning about Exploratory data analysis using pandas and air pollution and temperature data for the city of Chicago. 

Exploratory data analysis (EDA), pioneered by John Tukey, set a foundation for the field of data science. The key idea of EDA is that the first and most important step in any project based on data is to look at the data. By summarizing and visualizing the data, you can gain valuable intuition and understanding of the project.

The goal of exploratory data analysis is to get you thinking about your data and reasoning about your question. i.e to make sure we have the right data, identify any problems with the dataset, determining if what we answer our desired question and get rough idea of what the answer will look like.

Exploratory data analysis is the search for patterns and trends in a given data set. Looking at the bigger picture bigger picture Data-driven science, we start by collecting a data set of resonable size, and then looking for patterns that ideally will play the role of hypotheses for future analysis[2].  After the basic exploratory analysis we can pause and think that if our question needs refinement or if we need to collect more or new data.

Exploratory data analysis is about looking carefully at your dataset and identify any errors in data collection processing, finding violations of statistical assumptions, and suggesting interesting hypotheses[2].

As first step we could answer the following questions: Who constructed this dataset, when, and why? what is size of the data and the description of the various columns?


##### Topics Covered:

- Pandas functions for computing descriptive statistics
- Extracting rows based on conditionals 
- Sorting based on row values
- Compute relative difference 
- renaming columns, groupby function etc 


##### Acknowledgement 
topics and code inspired from book chapter 4: Exploratory Data Analysis with R
Text: 


In [ ]:
pip install pyreadr

     |████████████████████████████████| 251 kB 7.2 MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [ ]:
import pyreadr
import pandas as pd

In [ ]:
result = pyreadr.read_r('chicago')

In [ ]:
df = result[None]

##### Basic Stats using shape(), describe(),info() and summary ()

In [ ]:
df.shape


(6940, 8)

In [ ]:
df.describe()

tmpd         dptp   pm25tmean2   pm10tmean2     o3tmean2  \
count  6939.000000  6938.000000  2493.000000  6698.000000  6940.000000   
mean     50.309339    40.341686    16.230958    33.895206    19.435513   
std      19.412801    18.487240     8.697750    17.967363    11.385984   
min     -16.000000   -25.625000     1.700000     2.000000     0.152778   
25%      35.000000    27.000000     9.700000    21.500000    10.072917   
50%      51.000000    39.875000    14.657143    30.278846    18.521802   
75%      67.000000    55.750000    20.600000    42.000000    27.000996   
max      92.000000    78.250000    61.500000   365.000000    66.587500   

         no2tmean2  
count  6940.000000  
mean     25.231882  
std       7.991389  
min       6.158333  
25%      19.653819  
50%      24.555556  
75%      30.139040  
max      62.479984

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6940 entries, 0 to 6939
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   city        6940 non-null   object 
 1   tmpd        6939 non-null   float64
 2   dptp        6938 non-null   float64
 3   date        6940 non-null   object 
 4   pm25tmean2  2493 non-null   float64
 5   pm10tmean2  6698 non-null   float64
 6   o3tmean2    6940 non-null   float64
 7   no2tmean2   6940 non-null   float64
dtypes: float64(6), object(2)
memory usage: 433.9+ KB


In [ ]:
df.head()

city  tmpd    dptp        date  pm25tmean2  pm10tmean2  o3tmean2  no2tmean2
0  chic  31.5  31.500  1987-01-01         NaN   34.000000  4.250000  19.988095
1  chic  33.0  29.875  1987-01-02         NaN         NaN  3.304348  23.190994
2  chic  33.0  27.375  1987-01-03         NaN   34.166667  3.333333  23.815476
3  chic  29.0  28.625  1987-01-04         NaN   47.000000  4.375000  30.434524
4  chic  32.0  28.875  1987-01-05         NaN         NaN  4.750000  30.333333

In [ ]:
df_subset = df

In [ ]:
df_subset

city  tmpd    dptp        date  pm25tmean2  pm10tmean2  o3tmean2  \
0     chic  31.5  31.500  1987-01-01         NaN   34.000000  4.250000   
1     chic  33.0  29.875  1987-01-02         NaN         NaN  3.304348   
2     chic  33.0  27.375  1987-01-03         NaN   34.166667  3.333333   
3     chic  29.0  28.625  1987-01-04         NaN   47.000000  4.375000   
4     chic  32.0  28.875  1987-01-05         NaN         NaN  4.750000   
...    ...   ...     ...         ...         ...         ...       ...   
6935  chic  40.0  33.600  2005-12-27   23.560000   27.000000  4.468750   
6936  chic  37.0  34.500  2005-12-28   17.750000   27.500000  3.260417   
6937  chic  35.0  29.400  2005-12-29    7.450000   23.500000  6.794837   
6938  chic  36.0  31.000  2005-12-30   15.057143   19.200000  3.034420   
6939  chic  35.0  30.100  2005-12-31   15.000000   23.500000  2.531250   

      no2tmean2  
0     19.988095  
1     23.190994  
2     23.815476  
3     30.434524  
4     30.333333  
...         ...  
6935  23.500000  
6936  19.285628  
6937  19.972222  
6938  22.805556  
6939  13.250000  

[6940 rows x 8 columns]

#### Select a subset of 3 first columns


In [ ]:
df_subset[['city','tmpd','dptp']] 

city  tmpd    dptp
0     chic  31.5  31.500
1     chic  33.0  29.875
2     chic  33.0  27.375
3     chic  29.0  28.625
4     chic  32.0  28.875
...    ...   ...     ...
6935  chic  40.0  33.600
6936  chic  37.0  34.500
6937  chic  35.0  29.400
6938  chic  36.0  31.000
6939  chic  35.0  30.100

[6940 rows x 3 columns]

In [ ]:
df_filtered_rows = df.loc[df['pm25tmean2'] >30] #extract the rows of the chicago data frame where the levels of
#PM2.5 are greater than 30 (which is a reasonably high level)

In [ ]:
df_filtered_rows.head()

city  tmpd  dptp        date  pm25tmean2  pm10tmean2   o3tmean2  \
4034  chic  23.0  21.9  1998-01-17       38.10   32.461538   3.180556   
4040  chic  28.0  25.8  1998-01-23       33.95   38.692308   1.750000   
4137  chic  55.0  51.3  1998-04-30       39.40   34.000000  10.786232   
4138  chic  59.0  53.7  1998-05-01       35.40   28.500000  14.295125   
4139  chic  57.0  52.0  1998-05-02       33.30   35.000000  20.662879   

      no2tmean2  
4034  25.300000  
4040  29.376299  
4137  25.313095  
4138  31.429046  
4139  26.798611

In [ ]:
df_filtered_rows['pm25tmean2'].describe() 

count    194.000000
mean      36.626268
std        5.742143
min       30.050000
25%       32.125000
50%       35.042857
75%       39.533125
max       61.500000
Name: pm25tmean2, dtype: float64

there are now only 194 rows in the data frame and the distribution of
the pm25tmean2 values is as above

### Task: Extract the rows where PM2.5 is greater than 30 and temperature is greater than #80 degrees Fahrenheit.

In [ ]:
df_filtered_rows = df.loc[(df['pm25tmean2'] >30) & (df['tmpd'] > 80)]

In [ ]:
df_filtered_rows 

city  tmpd  dptp        date  pm25tmean2  pm10tmean2   o3tmean2  \
4252  chic  81.0  71.2  1998-08-23   39.600000        59.0  45.863636   
4266  chic  81.0  70.4  1998-09-06   31.500000        50.5  50.662500   
5314  chic  82.0  72.2  2001-07-20   32.300000        58.5  33.003804   
5326  chic  84.0  72.9  2001-08-01   43.700000        81.5  45.177355   
5333  chic  85.0  72.6  2001-08-08   38.837500        70.0  37.980468   
5334  chic  84.0  72.6  2001-08-09   38.200000        66.0  36.732452   
5649  chic  82.0  67.4  2002-06-20   33.000000        80.5  47.426731   
5652  chic  82.0  63.5  2002-06-23   42.500000        65.0  54.880435   
5667  chic  81.0  70.4  2002-07-08   33.100000        64.0  45.349693   
5677  chic  82.0  66.2  2002-07-18   38.850000        72.5  44.980455   
6019  chic  82.0  65.1  2003-06-25   33.900000        66.0  56.136662   
6028  chic  84.0  68.4  2003-07-04   32.900000        47.5  45.661458   
6749  chic  86.0  63.4  2005-06-24   31.857143        74.0  50.966486   
6752  chic  82.0  64.6  2005-06-27   51.537500        79.0  55.235857   
6753  chic  85.0  64.1  2005-06-28   31.200000        57.5  50.291440   
6772  chic  84.0  67.0  2005-07-17   32.700000        42.5  44.643229   
6789  chic  84.0  69.0  2005-08-03   37.900000        64.0  39.321105   

      no2tmean2  
4252  14.326389  
4266  20.312500  
5314  33.675000  
5326  27.442391  
5333  27.627433  
5334  26.467424  
5649  30.767029  
5652  30.039130  
5667  27.678571  
5677  26.069048  
6019  22.949337  
6028  21.343750  
6749  23.750000  
6752  28.549366  
6753  26.553977  
6772  16.270833  
6789  23.619318

In [ ]:
df_filtered_rows[['date','tmpd','pm25tmean2']] 

date  tmpd  pm25tmean2
4252  1998-08-23  81.0   39.600000
4266  1998-09-06  81.0   31.500000
5314  2001-07-20  82.0   32.300000
5326  2001-08-01  84.0   43.700000
5333  2001-08-08  85.0   38.837500
5334  2001-08-09  84.0   38.200000
5649  2002-06-20  82.0   33.000000
5652  2002-06-23  82.0   42.500000
5667  2002-07-08  81.0   33.100000
5677  2002-07-18  82.0   38.850000
6019  2003-06-25  82.0   33.900000
6028  2003-07-04  84.0   32.900000
6749  2005-06-24  86.0   31.857143
6752  2005-06-27  82.0   51.537500
6753  2005-06-28  85.0   31.200000
6772  2005-07-17  84.0   32.700000
6789  2005-08-03  84.0   37.900000

In [ ]:
df_filtered_rows['date'] = pd.to_datetime(df_filtered_rows.date)


<ipython-input-57-7e67a5daefa4>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered_rows['date'] = pd.to_datetime(df_filtered_rows.date)


In [ ]:
df_filtered_rows.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17 entries, 4252 to 6789
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   city        17 non-null     object        
 1   tmpd        17 non-null     float64       
 2   dptp        17 non-null     float64       
 3   date        17 non-null     datetime64[ns]
 4   pm25tmean2  17 non-null     float64       
 5   pm10tmean2  17 non-null     float64       
 6   o3tmean2    17 non-null     float64       
 7   no2tmean2   17 non-null     float64       
 8   Date        17 non-null     datetime64[ns]
dtypes: datetime64[ns](2), float64(6), object(1)
memory usage: 1.3+ KB


In [ ]:
df_sorted = df_filtered_rows[['date','pm25tmean2']].sort_values(by='date')


In [ ]:
df_sorted.head()

date  pm25tmean2
4252 1998-08-23     39.6000
4266 1998-09-06     31.5000
5314 2001-07-20     32.3000
5326 2001-08-01     43.7000
5333 2001-08-08     38.8375

In [ ]:
df_sorted.tail()

date  pm25tmean2
6749 2005-06-24   31.857143
6752 2005-06-27   51.537500
6753 2005-06-28   31.200000
6772 2005-07-17   32.700000
6789 2005-08-03   37.900000

### RENAME SPECIFIC COLUMNS


In [ ]:
df = df.rename(columns={'dptp': 'dewpoint', 'pm25tmean2': 'pm25'})

In [ ]:
df.head()

city  tmpd  dewpoint        date  pm25  pm10tmean2  o3tmean2  no2tmean2
0  chic  31.5    31.500  1987-01-01   NaN   34.000000  4.250000  19.988095
1  chic  33.0    29.875  1987-01-02   NaN         NaN  3.304348  23.190994
2  chic  33.0    27.375  1987-01-03   NaN   34.166667  3.333333  23.815476
3  chic  29.0    28.625  1987-01-04   NaN   47.000000  4.375000  30.434524
4  chic  32.0    28.875  1987-01-05   NaN         NaN  4.750000  30.333333

In [ ]:
# we create a pm25detrend variable that subtracts the mean from the pm25 variable

In [ ]:
df_new = df.assign(pm25detrend = df["pm25"] - df["pm25"].mean()) 

In [ ]:

df_new.tail()

city  tmpd  dewpoint        date       pm25  pm10tmean2  o3tmean2  \
6935  chic  40.0      33.6  2005-12-27  23.560000        27.0  4.468750   
6936  chic  37.0      34.5  2005-12-28  17.750000        27.5  3.260417   
6937  chic  35.0      29.4  2005-12-29   7.450000        23.5  6.794837   
6938  chic  36.0      31.0  2005-12-30  15.057143        19.2  3.034420   
6939  chic  35.0      30.1  2005-12-31  15.000000        23.5  2.531250   

      no2tmean2  pm25detrend  
6935  23.500000     7.329042  
6936  19.285628     1.519042  
6937  19.972222    -8.780958  
6938  22.805556    -1.173815  
6939  13.250000    -1.230958

In [ ]:
df['date'] = pd.to_datetime(df.date)



In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6940 entries, 0 to 6939
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   city        6940 non-null   object        
 1   tmpd        6939 non-null   float64       
 2   dewpoint    6938 non-null   float64       
 3   date        6940 non-null   datetime64[ns]
 4   pm25        2493 non-null   float64       
 5   pm10tmean2  6698 non-null   float64       
 6   o3tmean2    6940 non-null   float64       
 7   no2tmean2   6940 non-null   float64       
dtypes: datetime64[ns](1), float64(6), object(1)
memory usage: 433.9+ KB


In [ ]:
df['year'] = df['date'].dt.year


In [ ]:
df.head()

city  tmpd  dewpoint       date  pm25  pm10tmean2  o3tmean2  no2tmean2  \
0  chic  31.5    31.500 1987-01-01   NaN   34.000000  4.250000  19.988095   
1  chic  33.0    29.875 1987-01-02   NaN         NaN  3.304348  23.190994   
2  chic  33.0    27.375 1987-01-03   NaN   34.166667  3.333333  23.815476   
3  chic  29.0    28.625 1987-01-04   NaN   47.000000  4.375000  30.434524   
4  chic  32.0    28.875 1987-01-05   NaN         NaN  4.750000  30.333333   

   year  
0  1987  
1  1987  
2  1987  
3  1987  
4  1987

In [ ]:
df_new  = df.groupby("year", sort=True)

In [ ]:
df_new.head()

city  tmpd  dewpoint       date     pm25  pm10tmean2   o3tmean2  \
0     chic  31.5    31.500 1987-01-01      NaN   34.000000   4.250000   
1     chic  33.0    29.875 1987-01-02      NaN         NaN   3.304348   
2     chic  33.0    27.375 1987-01-03      NaN   34.166667   3.333333   
3     chic  29.0    28.625 1987-01-04      NaN   47.000000   4.375000   
4     chic  32.0    28.875 1987-01-05      NaN         NaN   4.750000   
...    ...   ...       ...        ...      ...         ...        ...   
6575  chic  33.0    24.800 2005-01-01  10.5250   16.500000   9.208333   
6576  chic  44.0    39.600 2005-01-02      NaN   16.000000   6.447917   
6577  chic  34.0    31.500 2005-01-03  17.5000   11.500000  12.687500   
6578  chic  33.0    29.000 2005-01-04   8.8375   11.600000  17.548460   
6579  chic  29.0    25.500 2005-01-05   3.3000    9.000000  27.760417   

      no2tmean2  year  
0     19.988095  1987  
1     23.190994  1987  
2     23.815476  1987  
3     30.434524  1987  
4     30.333333  1987  
...         ...   ...  
6575  14.152778  2005  
6576  16.069444  2005  
6577  18.625845  2005  
6578  19.315217  2005  
6579  13.753788  2005  

[95 rows x 9 columns]